### PR TITLE
New specs for a python3 wmagent and wmcore-devtools

### PR DESCRIPTION
**Pull request by amaltaro**
This is a work in progress for building WMAgent in a python3 environment.

### CODE REVIEW

**Comment by amaltaro** [comment]
Among a bunch of other errors, the first that caught my attention was the execution of a py2 init.sh script before running pip3 to download a package, as seen in [1]. Short version is:
```
. /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pylint==1.9.4;
```

@smuzaffar Shahzad, perhaps it's a known issue for you and you can point me in the right direction(?)

[1]
```
Package external+py2-pip+9.0.3-comp2 found in repository
Starting to process package external+py3-setuptools+39.2.0
Checking if py3-setuptools is cached.
Download command: cd /build/amaltaro/w/tmp/tmpQAwMmg; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pep8==1.7.0; /bin/ls | grep -v files.list > files.list
Download command: cd /build/amaltaro/w/tmp/tmpWwuv4z; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . CherryPy==17.4.0; /bin/ls | grep -v files.list > files.list
Package external+py3-setuptools+39.2.0 not found in repository. Queuing for build.
Download command: cd /build/amaltaro/w/tmp/tmp8_1Nk7; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pylint==1.9.4; /bin/ls | grep -v files.list > files.list
Download command: cd /build/amaltaro/w/tmp/tmpzInU4M; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . nose==1.3.7; /bin/ls | grep -v files.list > files.list
Dependencies for external+py3-setuptools+39.2.0: ['install-external+gcc+6.3.0', 'install-external+python3+3.6.2', 'fetch-external+py3-setuptools+39.2.0']
Error while downloading sources from pip://pep8/1.7.0?pip_options=--no-deps%20--no-binary%3D:all:&pip=pip3&output=/source.tar.gz.

cd /build/amaltaro/w/tmp/tmpQAwMmg; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download  --no-deps --no-binary=:all: --disable-pip-version-check  -d . pep8==1.7.0; /bin/ls | grep -v files.list > files.list 
256 
Traceback (most recent call last):
File "/usr/bin/pip3", line 8, in <module>
from pip import main
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/__init__.py", line 43, in <module>
from pip.utils import get_installed_distributions, get_prog
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/utils/__init__.py", line 27, in <module>
from pip._vendor import pkg_resources
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 35, in <module>
import plistlib
File "/usr/lib64/python3.6/plistlib.py", line 65, in <module>
from xml.parsers.expat import ParserCreate
File "/usr/lib64/python3.6/xml/parsers/expat.py", line 4, in <module>
from pyexpat import *
ImportError: /usr/lib64/python3.6/lib-dynload/pyexpat.cpython-36m-x86_64-linux-gnu.so: undefined symbol: XML_SetHashSalt

During handling of the above exception, another exception occurred:
```

**Comment by muhammadimranfarooqi** [comment]
-1 to 174cc41497b0895c094ebc0f8dec8d6e79de3092 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_174cc41-20200614-140503.log" target="_blank">here</a>.

**Comment by smuzaffar** [comment]
@amaltaro ,which branch of pkgtools you are using?

**Comment by amaltaro** [comment]
This is my pkgtools head:
```
9da43d7e09c208f66e44d8a6a081451f3415cce5 (HEAD, V00-32-XX) do not delete rpm on failure if -k option is set
```

**Comment by smuzaffar** [comment]
by the way, cmsBuild does not support two different packages of pip. I would suggest to get py2-pip.spec from IB/CMSSW_11_2_X/master branch where pip is build for both py2 and py3 in same package.

**Comment by smuzaffar** [comment]
@amaltaro , did you try getting py2-pip.spec from IB/CMSSW_11_2_X/master branch?

**Comment by amaltaro** [comment]
@smuzaffar I see that py2-pip also depends and builds python3. One of the issues we would have is related to the new environment variables set by that spec (PYTHON27PATH / PYTHON3PATH), which we would have to update everything else in the comp_gcc630 branch.

The approach I'm taking here is to actually create a new spec file for wmagent, such that we can decide whether we build the spec based on python3 or python2. So we have also created a pip3 based build-with-pip:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/build-with-pip3.file#L30

**Comment by smuzaffar** [comment]
are you dropping python2 completely? If you want to build full stack with python3 then that is fine. In that case you need to pass [a] `--pip-package py3-pip` to cmsBuild. Note that in this case your full stack should use python3.

[a]
https://github.com/cms-sw/pkgtools/blob/V00-32-XX/cmsBuild#L2798

**Comment by amaltaro** [comment]
@smuzaffar I replied to you via mattermost because it might be quicker to discuss it.
In short, we cannot update the whole stack because we will start gradually migrating services (actually, keeping services compatible with both python versions), this is why I also decided to go with a new spec for wmagent (wmagentpy3).

**Comment by smuzaffar** [comment]
@amaltaro , please try to use latest pkgtools V00-32-Xx branch. It supports pip package name via source url. For example in your `build-with-pip3.file` https://github.com/cms-sw/cmsdist/blob/174cc41497b0895c094ebc0f8dec8d6e79de3092/build-with-pip3.file#L34 you can have 

```
%define source0 pip://%{pip_name}/%{realversion}?pip_options=%{PipDownloadOptions}&pip=pip3&pip_package=py3-pip&output=/%{pkgsource}
```
i.e. add `pip_package=py3-pip&`. This will instruct cmsBuild to depend on py3-pip instead of py2-pip.

**Comment by muhammadimranfarooqi** [comment]
-1 to fac8294841a55d5eca09696c1bc9e36722976f7f on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_fac8294-20200617-083004.log" target="_blank">here</a>.

**Comment by smuzaffar** [comment]
this needs latest pkgtools V00-32-XX branch

**Comment by amaltaro** [comment]
@smuzaffar in general, it seems to be properly building python3 packages now with the new pkgtools and updated build-with-pip3 files.

However, it's the second time my build gets stuck. This is the command I'm executing:
```
pkgtools/cmsBuild -c cmsdist --repository comp \
>    -a slc7_amd64_gcc630 --builders 8 -j 4 --work-dir w \
>    build comp | tee logBuild
```

and the first time it got stuck (for a couple of hours), this was the tail of the terminal output:
```
Done external+py3-requests+2.18.4
Build successful py3-docutils.
(<function buildPackage at 0x7f92431291b8>, <Package name=py3-docutils>, <scheduler.Scheduler object at 0x7f9242c540d0>) done
Trying to install the rpm package external+py3-docutils+0.12 just built.
Checking local path dependency for rpm package external+py3-docutils+0.12 just build.
Done external+py3-docutils+0.12
```

I stopped and started running again. It's now stuck here:
```
Building py3-mysqldb. Log can be found in /build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/py3-mysqldb/1.2.4b4/log.
rm -rf /build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/py3-mysqldb/1.2.4b4/pkgtools-pkg-src-move2git ; export _CMSBUILD_BUILD_ENV_=1;  source /build/amaltaro/w/slc7_amd64_gcc630/rpm-env.sh ;  TMPDIR=/build/amaltaro/w/tmp  rpmbuild  --buildroot /build/amaltaro/w/tmp/BUILDROOT/d07fe2fe39af9efcf3c10c30f761a5d5 -ba --define '_topdir /build/amaltaro/w' --define "compiling_processes 4"  --define "compilerv 630" --define "cmscompilerv 630" --define "cmsos slc7_amd64" /build/amaltaro/w/SPECS/external/py3-mysqldb/1.2.4b4/spec >/build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/py3-mysqldb/1.2.4b4/log 2>&1
(<function buildPackage at 0x7f85603be1b8>, <Package name=py3-mysqldb>, <scheduler.Scheduler object at 0x7f855fdcfa50>) done
```

is there any magic that I can do to check what's going on? It's on vocms055 btw

**Comment by smuzaffar** [comment]
try running with --debug and --trace options

**Comment by smuzaffar** [comment]
while you are at it, may be you can update python3 to match the version from cmssw (we are using 3.8.2 )

**Comment by amaltaro** [comment]
Yes, upgrading python is on my todo list. I just wanted to isolate these changes for the moment to make sure that the new python3 version will not affect the current changes I have.

About the debug/trace options, I'm afraid it does not help me. Building got stuck at the same stage. Just in case, this is the command I executed:
```
amaltaro@vocms055:/build/amaltaro $ pkgtools/cmsBuild -c cmsdist --repository comp    -a slc7_amd64_gcc630 --builders 8 -j 4 --work-dir w    build comp --debug --trace | tee logBuild 
The following commands will be executed:
* bootstrap
* build
rm -rf /build/amaltaro/w/tmp/BUILDROOT

gcc has checksum 269419e7dcbe2ff4333a4c6075dfa523
Matching Packages all 269419e7dcbe2ff4333a4c6075dfa523: external+gcc+6.3.0,external+gcc+6.3.0
269419e7dcbe2ff4333a4c6075dfa523 is aliased to '', using ''
...
```

**Comment by amaltaro** [comment]
Ohh, checking the py3-mysqldb log, I see the build has failed with:
```
+ python3 setup.py build
Traceback (most recent call last):
  File "setup.py", line 13, in <module>
    from setup_posix import get_config
  File "/build/amaltaro/w/BUILD/slc7_amd64_gcc630/external/py3-mysqldb/1.2.4b4/MySQL-python-1.2.4b4/setup_posix.py", line 2, in <module>
    from ConfigParser import SafeConfigParser
ModuleNotFoundError: No module named 'ConfigParser'
error: Bad exit status from /build/amaltaro/w/tmp/rpm-tmp.Zh05za (%build)
```

**Comment by muhammadimranfarooqi** [comment]
-1 to b40e2481de203515bb68939198ec2abdcbcb2ef2 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_b40e248-20200617-171003.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
So, I tried different combinations of parameters (even with a single builder and process) and still it gets stuck at some point (last 3 cases it was in the py3-six spec).
I'll add some extra debugging tomorrow to see if I can spot where the issue is (--debug/--trace aren't verbose enough for this).

**Comment by smuzaffar** [comment]
I would suggest to build with `--builder 1` to see which package is stuck building

**Comment by smuzaffar** [comment]
@amaltaro wait please, `pip_package=name` is not fully working, let me fix that first

**Comment by smuzaffar** [comment]
@amaltaro , please try using pkgtools branch v32-pip-pkg which I am testing here https://github.com/cms-sw/pkgtools/pull/158 . Once tests run fine then I will merge it in to V00-32-XX branch.

I noticed `cmsBuild` already supported `package_dependency=<packages>` for source url. So instead of adding a new `pip_package=` option, now I have updated cmsBuild to make use of `package_dependency` for `pip://` urls  too. So please use `package_dependency=py3-pip` in the build-with-pip4.file.

**Comment by amaltaro** [comment]
Thanks @smuzaffar . I just started building packages. Will let you know how it goes in the coming hour.

**Comment by amaltaro** [comment]
I'm seeing some errors in the build, but it could be because I'm trying to pull pypi packages that do do not support python3.6.

Just to make sure I understood you, by `package_dependency=py3-pip`, did you mean this line:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/build-with-pip3.file#L31
?

I'm also considering to actually update the python version now, at least I suffer only once... :)

**Comment by amaltaro** [comment]
I updated the `mock` package version and I still see an error like:
```
Package external+py3-future+0.16.0 not found in repository. Queuing for build.
Dependencies for external+py3-future+0.16.0: ['fetch-external+py3-future+0.16.0']
Download command: cd /build/amaltaro/w/tmp/tmpR5LYnx; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . mock==4.0.2; /bin/ls | grep -v files.list > files.list
Error while downloading sources from pip://mock/4.0.2?pip_options=--no-deps%20--no-binary%3D:all:&pip=pip3&pip_package=py3-pip&output=/source.tar.gz.

cd /build/amaltaro/w/tmp/tmpR5LYnx; . /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh; pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . mock==4.0.2; /bin/ls | grep -v files.list > files.list 
256 
Traceback (most recent call last):
File "/usr/bin/pip3", line 8, in <module>
from pip import main
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/__init__.py", line 43, in <module>
from pip.utils import get_installed_distributions, get_prog
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/utils/__init__.py", line 27, in <module>
from pip._vendor import pkg_resources
File "/build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/lib/python2.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 35, in <module>
import plistlib
File "/usr/lib64/python3.6/plistlib.py", line 65, in <module>
from xml.parsers.expat import ParserCreate
File "/usr/lib64/python3.6/xml/parsers/expat.py", line 4, in <module>
from pyexpat import *
ImportError: /usr/lib64/python3.6/lib-dynload/pyexpat.cpython-36m-x86_64-linux-gnu.so: undefined symbol: XML_SetHashSalt
```

This mock==4.0.2 is meant to support py3.2+

Another problem is that I keep seeing the pip2 env being used:
```
. /build/amaltaro/w/slc7_amd64_gcc630/external/py2-pip/9.0.3-comp2/etc/profile.d/init.sh
```

**Comment by smuzaffar** [comment]
you are still using `pip_package=py3-pip` , please change it to `package_dependency=py3-pip`

**Comment by amaltaro** [comment]
Gotcha! So far so good!

**Comment by muhammadimranfarooqi** [comment]
-1 to 16ca97eed2fb727760432e7a9df7dc9788358c50 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_16ca97e-20200618-150503.log" target="_blank">here</a>.

**Comment by smuzaffar** [comment]
@amaltaro , pkgtools changes look good to me. I was able to build py3-moch from your branch and cms bot tests are also +ve. so I have merged pkgtools changes. V00-32-XX branch should be good enough foryour py3 changes,

**Comment by muhammadimranfarooqi** [comment]
-1 to 4fc428d8b6e47a30273c28225334361d0a84c21a on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_4fc428d-20200618-205004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to 56cc88dd81dbb3a8015b3f07873fd21468f4f9c7 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_56cc88d-20200618-215504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200618-225504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-000220.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-000505.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Thanks Shahzad. I managed to build a bunch of packages, had to fix and create many others (including a bunch of crap libraries that others depend on). 

Now I'm hitting another wall building py3-extras. Every time I run cmsBuild, it crashes like this:
```
Generated cache: external+py3-pbr+1.8.1
Generated cache: external+py3-extras+1.0.0
Traceback (most recent call last):
  File "pkgtools/cmsBuild", line 4307, in <module>
    build(opts, args[1:], PKGFactory)
  File "pkgtools/cmsBuild", line 3630, in build
    packages = [factory.createWithSpec(pkgName) for pkgName in args]
  File "pkgtools/cmsBuild", line 1419, in createWithSpec
    pkg.initWithSpec(specLines)
...
  File "pkgtools/cmsBuild", line 1162, in __getitem__
    return self.__cache.__getitem__(self.__decorator(name))
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

this extras is one of the dependencies of testtools, which is a dependency of fixtures, required by mox3... 
I will try to build things without mox3 tomorrow, but in case you know why this RuntimeError is popping up, please do let me know.

**Comment by smuzaffar** [comment]
looks like a cyclic dependency to me

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-001005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-001505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-002005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-002504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-003005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-003505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-004004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-004505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-005005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-005504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-010505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-011005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-011505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-012005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-012505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-013005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-013505.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-014004.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-014504.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-015005.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to c818d1212ffa2e3b4f77c35542ce670d833c7d5c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/5939/test_c818d12-20200619-015505.log" target="_blank">here</a>.

**Comment by smuzaffar** [comment]
Closing it in hope that dmwm bot will stop testing

**Comment by smuzaffar** [comment]
@amaltaro , I had to close this to stop dmwm bot. were you able to fixgure out the reason of your last error? It looks like a cyclic deps issue to me

**Comment by amaltaro** [comment]
Ohh, it looks like I missed all these notifications. Thanks for closing this issue. I have already checked for cyclic dependencies and could not find anything. But I'll get back to this later today.

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, I added a few debugging statements to the cmsBuild script, and it seems to hop back and forth on two python spec files, see the logs:
```
Read from cache: external+py3-isort+4.3.17
AMR dealing with: py2-scipy
Read from cache: external+py2-scipy+1.1.0
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
AMR dealing with: py3-testtools
AMR self.origRequires: ['py3-extras', 'py3-fixtures', 'py3-pbr', 'py3-python-mimeparse', 'py3-six', 'py3-traceback2', 'py3-unittest2', 'python3', 'py3-setuptools']
AMR dealing with: py3-extras
Read from cache: external+py3-extras+1.0.0
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
AMR dealing with: py3-testtools
AMR self.origRequires: ['py3-extras', 'py3-fixtures', 'py3-pbr', 'py3-python-mimeparse', 'py3-six', 'py3-traceback2', 'py3-unittest2', 'python3', 'py3-setuptools']
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
AMR dealing with: py3-testtools
AMR self.origRequires: ['py3-extras', 'py3-fixtures', 'py3-pbr', 'py3-python-mimeparse', 'py3-six', 'py3-traceback2', 'py3-unittest2', 'python3', 'py3-setuptools']
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
AMR dealing with: py3-testtools
AMR self.origRequires: ['py3-extras', 'py3-fixtures', 'py3-pbr', 'py3-python-mimeparse', 'py3-six', 'py3-traceback2', 'py3-unittest2', 'python3', 'py3-setuptools']
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
AMR dealing with: py3-testtools
AMR self.origRequires: ['py3-extras', 'py3-fixtures', 'py3-pbr', 'py3-python-mimeparse', 'py3-six', 'py3-traceback2', 'py3-unittest2', 'python3', 'py3-setuptools']
AMR dealing with: py3-fixtures
AMR self.origRequires: ['py3-pbr', 'py3-six', 'py3-testtools', 'python3', 'py3-setuptools']
....
```

Looking at the `fixtures` and `testtools`, I don't see any circular dependencies. Do you see anything wrong there?

File: py3-testtools.spec [hunk 1]
```diff
@@ -0,0 +1,5 @@
+### RPM external py3-testtools 2.4.0
+## IMPORT build-with-pip3
+
+Requires: py3-extras py3-fixtures py3-pbr py3-python-mimeparse py3-six py3-traceback2 py3-unittest2
```
Comment (davidlange6) [inline py3-testtools.spec:4]: seems the circular dependency is clear - here testtbools wants fixtures

File: py3-fixtures.spec [hunk 2]
```diff
@@ -0,0 +1,5 @@
+### RPM external py3-fixtures 3.0.0
+## IMPORT build-with-pip3
+
+Requires: py3-pbr py3-six py3-testtools
```
Comment (davidlange6) [inline py3-fixtures.spec:4]: and here fixtures wants testtools

**Comment by amaltaro** [comment]
@davidlange6 I'm sorry, you're right!

What would be the recommended way to deal with it? Dropping one of those specs?

**Comment by davidlange6** [comment]
Presumably one of the two is not correctly listing it’s dependencies



On 1 Jul 2020, at 08:33, Alan Malta Rodrigues <notifications@github.com> wrote:

﻿

@davidlange6<https://github.com/davidlange6> I'm sorry, you're right!

What would be the recommended way to deal with it? Dropping one of those specs?

**Comment by amaltaro** [comment]
I have made some further changes to my local build in vocms055, and indeed cmsBuild complains about missing dependencies for both packages. Example for `testtools`:
```
++ pip3 show testtools --disable-pip-version-check
++ grep '^Name:'
++ sed 's|^Name: *||;s| ||g'
+ PKG_NAME=testtools
++ pip3 check --disable-pip-version-check
++ grep '^testtools  *2.4.0  *requires '
++ sed 's|,.*||;s|.* |py3-|'
++ tr '\n' ' '
+ DEPS='py3-fixtures py3-unittest2 '
+ '[' 'py3-fixtures py3-unittest2 ' '!=' '' ']'
+ echo 'ERROR: Missing dependencies for py3-testtools (python2) found: py3-fixtures py3-unittest2 '
ERROR: Missing dependencies for py3-testtools (python2) found: py3-fixtures py3-unittest2 
+ exit 1
error: Bad exit status from /build/amaltaro/w/tmp/rpm-tmp.YclCN3 (%build)
```

and a few cmsBuild build runs before this, I've got complaints to the `fixtures` package too:
```
++ pip3 show fixtures --disable-pip-version-check
++ grep '^Name:'
++ sed 's|^Name: *||;s| ||g'
+ PKG_NAME=fixtures
++ pip3 check --disable-pip-version-check
++ grep '^fixtures  *3.0.0  *requires '
++ sed 's|,.*||;s|.* |py3-|'
++ tr '\n' ' '
+ DEPS='py3-testtools '
+ '[' 'py3-testtools ' '!=' '' ']'
+ echo 'ERROR: Missing dependencies for py3-fixtures (python2) found: py3-testtools '
ERROR: Missing dependencies for py3-fixtures (python2) found: py3-testtools 
+ exit 1
```

Googling for the requirements of those packages, indeed cmsBuild is reporting the correct list of dependencies:
https://bazaar.launchpad.net/~testing-cabal/python-fixtures/master/view/head:/requirements.txt
and
https://github.com/testing-cabal/testtools/blob/master/requirements.txt

So, I guess the only option left is to actually try to replace one of those by a different library(?) It will be painful though, because it means a bunch of other inner dependencies need to be properly dropped as well.

**Comment by smuzaffar** [comment]
is this PR up to date? I can try to work around this issue.

**Comment by amaltaro** [comment]
That would be extremely helpful! Thanks, Shahzad.
I have just updated the code with the changes I had in my local area.

Just to confirm as well, this is the pkgtools tag I've been using:
```
amaltaro@vocms055:/build/amaltaro/pkgtools $ git log -10 --pretty=oneline --decorate
89c2f759c4c7d356807ac817bb0bed359ed3d649 (HEAD, origin/V00-32-XX, V00-32-XX) Merge pull request #162 from mrodozov/json-deps-file
```

**Comment by amaltaro** [comment]
Actually I don't seem my latest commit:
```
e90b295fd57a4ed236f59b7025978ec28f325f50 (HEAD -> wma-py3) fix other python specs; added linecache2 spec
a7e8421389ec30a8cdfb61f5f7453be7182d1f67 added testtools dependencies; new specs
```

I wonder if it's because the PR is closed(?)

I do see it in my branch though:
https://github.com/amaltaro/cmsdist/commits/wma-py3

**Comment by smuzaffar** [comment]
this cyclic deps is known issue 

https://github.com/testing-cabal/testtools/issues/291 
https://github.com/python-poetry/poetry/issues/62

as it is only runtime dependency , so we can patch one of these to not require the other. This way we can build and install. All you need to make sure that any package using one of these should also use the other too (otherwise runtime might break).

I have opened a PR https://github.com/amaltaro/cmsdist/pull/1 to yopur branch which should break this deps.

**Comment by amaltaro** [comment]
@muhammadimranfarooqi  Imran, could you please update the pkgtools branch used by the cms-bot to build/test our COMP RPMs to the head of V00-32-XX?

Hopefully I will be able to reopen this PR later today. Please ping me on gtalk if you need further details.

**Comment by amaltaro** [comment]
@smuzaffar Shahzad, are you able to reopen this PR? I don't see the reopen button; it seems to be disabled for Imran as well.

**Comment by smuzaffar** [comment]
I would suggest to open a new PR.
"Reopen" button is disabled tell me that the branch was forced push or  recreated.

**Comment by amaltaro** [comment]
#6021 has been created. Thank you for the very valuable support so far ;)

**Pull request closed without merging**